File: docs/perry-main-src/perry-main/benchmarks/json_polyglot/bench_field_access.go
// JSON parse-and-iterate polyglot benchmark — Go.
// 10k records, ~1 MB blob, 50 iterations.
// Per iteration: parse → sum every record's nested.x → stringify.
// Identical workload to bench_field_access.ts/.rs/.swift/.cpp/.kt.

package main

import (
	"encoding/json"
	"fmt"
	"strconv"
	"time"
)

type FANested struct {
	X int `json:"x"`
	Y int `json:"y"`
}

type FAItem struct {
	Id     int      `json:"id"`
	Name   string   `json:"name"`
	Value  float64  `json:"value"`
	Tags   []string `json:"tags"`
	Nested FANested `json:"nested"`
}

func main() {
	items := make([]FAItem, 10000)
	for i := 0; i < 10000; i++ {
		items[i] = FAItem{
			Id:    i,
			Name:  "item_" + strconv.Itoa(i),
			Value: float64(i) * 3.14159,
			Tags:  []string{"tag_" + strconv.Itoa(i%10), "tag_" + strconv.Itoa(i%5)},
			Nested: FANested{X: i, Y: i * 2},
		}
	}
	blob, _ := json.Marshal(items)

	// Warmup
	for i := 0; i < 3; i++ {
		var parsed []FAItem
		_ = json.Unmarshal(blob, &parsed)
		warmSum := 0
		for j := range parsed {
			warmSum += parsed[j].Nested.X
		}
		_, _ = json.Marshal(parsed)
		_ = warmSum
	}

	const iterations = 50
	start := time.Now()

	checksum := 0
	for iter := 0; iter < iterations; iter++ {
		var parsed []FAItem
		_ = json.Unmarshal(blob, &parsed)
		sum := 0
		for i := range parsed {
			sum += parsed[i].Nested.X
		}
		checksum += sum
		reStringified, _ := json.Marshal(parsed)
		checksum += len(reStringified)
	}

	elapsed := time.Since(start).Milliseconds()
	fmt.Printf("ms:%d\n", elapsed)
	fmt.Printf("checksum:%d\n", checksum)
}

File: docs/perry-main-src/perry-main/benchmarks/json_polyglot/bench_field_access.kt
// JSON parse-and-iterate polyglot benchmark — Kotlin (kotlinx.serialization).
// 10k records, ~1 MB blob, 50 iterations.
// Per iteration: parse → sum every record's nested.x → stringify.
// Identical workload to bench_field_access.ts/.go/.rs/.swift/.cpp.

import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json

@Serializable
data class FANested(val x: Int, val y: Int)

@Serializable
data class FAItem(
    val id: Int,
    val name: String,
    val value: Double,
    val tags: List<String>,
    val nested: FANested,
)

fun main() {
    val items = (0 until 10_000).map { i ->
        FAItem(
            id = i,
            name = "item_$i",
            value = i * 3.14159,
            tags = listOf("tag_${i % 10}", "tag_${i % 5}"),
            nested = FANested(x = i, y = i * 2),
        )
    }
    val blob: String = Json.encodeToString(items)

    // Warmup
    repeat(3) {
        val parsed = Json.decodeFromString<List<FAItem>>(blob)
        var warmSum = 0L
        for (item in parsed) {
            warmSum += item.nested.x.toLong()
        }
        Json.encodeToString(parsed)
    }

    val iterations = 50
    val start = System.currentTimeMillis()

    var checksum = 0L
    repeat(iterations) {
        val parsed = Json.decodeFromString<List<FAItem>>(blob)
        var sum = 0L
        for (item in parsed) {
            sum += item.nested.x.toLong()
        }
        checksum += sum
        val reStringified = Json.encodeToString(parsed)
        checksum += reStringified.length.toLong()
    }

    val elapsed = System.currentTimeMillis() - start
    println("ms:$elapsed")
    println("checksum:$checksum")
}
